### ISSUE TITLE
(spotify) Install hangs if Windows store version of the application is present

### ISSUE THREAD

**Issue by splttingatms** [chocolatey-community/chocolatey-packages#941]
# problem: 
spotify installation using command `choco install spotify` hangs after showing the following message:
>SUCCESS: The scheduled task "spotify" has successfully been created.
>SUCCESS: Attempted to run the scheduled task "spotify".
>SUCCESS: The scheduled task "spotify" was successfully 

# cause: 
spotify's installation script has a do-until loop that waits for spotify process to start, mentioned below. The problem is that the spotify process never starts and so this do-until loop continues indefinitely.
https://github.com/chocolatey/chocolatey-coreteampackages/blob/80eb8095cf74f90ddae26e27205ca82d10a60da3/automatic/spotify/tools/ChocolateyInstall.ps1#L24-L33

I forked the package and enabled powershell tracing by adding the line `Set-PSDebug -Trace 2` to the top of the `ChocolateyInstall.ps1` script. I am not sure if this was what actually triggered this new window but after attempting to install my local package `choco install .\spotify.1.0.66.478.nupkg -dv -s "'$pwd;ht
tps://chocolatey.org/api/v2/'"` I saw the error window below. This error window was not showing before, only until I added the powershell tracing.

![chocolatey-spotify-error](https://user-images.githubusercontent.com/3055505/34375010-230f7356-ea99-11e7-818a-5aa503b178d6.png)

It turns out that the spotify process was never starting because the installer was hanging due my computer already having a spotify "app" installed. The powershell tracing shows that the do-until loop was executing in an infinite loop:
```
 28+  >>>> $done = $false
 30+   if ( >>>> Get-Process Spotify -ErrorAction SilentlyContinue) {
 30+   if ( >>>> Get-Process Spotify -ErrorAction SilentlyContinue) {
 35+    >>>> Start-Sleep -s 10
 30+   if ( >>>> Get-Process Spotify -ErrorAction SilentlyContinue) {
 30+   if ( >>>> Get-Process Spotify -ErrorAction SilentlyContinue) {
 35+    >>>> Start-Sleep -s 10
...
```

After uninstalling the Windows Store app version of spotify, I was able to re-run the install command successfully.

# solutions:
For end users, make sure to **uninstall Spotify Windows Store app** before installing the Spotify program version using chocolatey.

To fix script, a couple improvements can be made:
* include a pre-check for spotify app and prompt user of the conflict
* enable a way for spotify installer error window to appear in normal usage
* detect if installer task ends with error code and stop choco install if error detected
* add limit to do-until loop to prevent infinite loop in error case

related issue: #771

**Comment by AdmiringWorm** [comment]
Just quoting a comment on the disqus channel, which is probably something we should consider.

> posted by disqus user **damien**
> why not use runas instead of the scheduled task ?
>
> You can run an non privilege process from a provilege session with runas /trustlevel:0x20000
>
> To run the installer the command line would be something like: run /trustlevel:0x20000 "SpotifyFullSetup.exe /silent"

**Comment by majkinetor** [comment]
Interesting. Is this confirmed to work ? We used schtask on few places most notably virtualbox

/cc @ferventcoder

**Comment by majkinetor** [comment]
@AdmiringWorm 

I don't see how this privilege thing will help. It just has to fail after 2-3 iterations of the loop.

> For end users, make sure to uninstall Spotify Windows Store app before installing the Spotify program version using chocolatey.

This is definitely not something we should do as intrusion is massive. We throw in situations like this and let the user uninstall manually. 

So, this package has to be fixed to:

* Has limit on loop
* Can recognize above message or recognize spotify installation some other way.  Note that this should work on Windows 7 also which doesn't have store.

**Comment by AdmiringWorm** [comment]
> I don't see how this privilege thing will help. It just has to fail after 2-3 iterations of the loop.

of course, that's also an option, I'm just saying we should consider using runas. Or at least test it, there has been several reports that the spotify fails to install from them (can't remember the reasons though).

> This is definitely not something we should do as intrusion is massive. We throw in situations like this and let the user uninstall manually.

Yeah, that should be ignored, that's just a big no no. While we could warn users, just uninstalling the Windows Store app we really shouldn't do.

**Comment by jzajac2** [comment]
Well I do NOT have the Windows Store version installed (fresh off the press win10 install). That is, if I open Windows Store, search for Spotify, it says 'Get' (so I don't have it, right?)

> spotify v1.0.80.474 [Approved]
spotify package files install completed. Performing other installation steps.
File appears to be downloaded already. Verifying with package checksum to determine if it needs to be redownloaded.
Hashes match.
Hashes match.
SUCCESS: The scheduled task "spotify" has successfully been created.
SUCCESS: Attempted to run the scheduled task "spotify".
SUCCESS: The scheduled task "spotify" was successfully deleted.

**Comment by majkinetor** [comment]
I think this should be solved by throwing error in the package if the app is detected in Windows store, on those systems that have it.  Then user can decide what to do, uninstall the store app and repeat the choco install or keep using the store version.

**Comment by gunnsth** [comment]
Happens for me on fully updated Windows 10 instance (10.0.17763 Build 17763) with choco v0.10.11.

**Comment by AdmiringWorm** [comment]
@majkinetor do you know how to detect if it was installed from the store app?

**Comment by majkinetor** [comment]
I don't use store apps, even worst, I disabled the complete engine and turned off AppXSvc as it is a resource hog without obvious reason. 

Detecting store apps is trivial, but this one would have to be done by somebody that actually use this (I do not). Furthermore, there is that comment above of the occurrence of this problem even without store app. 

I personally think it doesn't justify effort and that this is upstream problem.

**Comment by majkinetor** [comment]
I looked at the 

> To run the installer the command line would be something like: run /trustlevel:0x20000 "SpotifyFullSetup.exe /silent"

This forum post claims that this shouldn't be used:
https://social.technet.microsoft.com/Forums/windows/en-US/cd82780b-6afa-4404-8515-5d646a5623de/quotrunasquot-command-trustlevel-what-is-it-how-can-i-enable-additional-levels

More info:
https://blogs.msdn.microsoft.com/aaron_margosis/2009/06/06/faq-how-do-i-start-a-program-as-the-desktop-user-from-an-elevated-app/

**Comment by mkevenaar** [comment]
I just had the same issue on my new laptop and I think we have 2 issues here:

1. If Spotify is installed using Windows Store
2. If the installation failed.

The current setup uses a scheduled task to start the installation. (see https://github.com/chocolatey-community/chocolatey-coreteampackages/blob/master/automatic/spotify/tools/ChocolateyInstall.ps1#L18-L22)

Then it sits and waits until the Spotify tool is started and it kills it. (see https://github.com/chocolatey-community/chocolatey-coreteampackages/blob/master/automatic/spotify/tools/ChocolateyInstall.ps1#L24-L33) 

But in the above cases, Spotify is not started and the script will wait until the cows come home.

Based on the script, there is no way we currently check that the installation was successful. I think we might need to refactor the installation script of Spotify.

**Comment by nickbeaugie** [comment]
I can confirm that this issue is still live. And I did not have the Store version installed prevously!

My install hangs after outputting these:
```
SUCCESS: The scheduled task "spotify" has successfully been created.
SUCCESS: Attempted to run the scheduled task "spotify".
SUCCESS: The scheduled task "spotify" was successfully deleted.
```

But it's not big deal as I'm able to install it via the Windows store. I think this is the same version now as if you go via Spotify's MSI anyway,

**Comment by eyadsibai** [comment]
I don't have the windows store version installed and still not able to install the choco version

**Comment by sbissonnette** [comment]
This is still an issue!

**Comment by majkinetor** [comment]
It is closed due to inactivity. No amount of reporting will change that. Unless someone contributes a solution or valuable info toward it, there is no need to reopen this.

**Comment by TSchnabl** [comment]
It is closed due to inactivity. No amount of reporting will change that. A contributor needs to provide a solution. Until then, nobody will care. Remember: This is not an endcustomer solution, chocolatey is only for people who know that they have to solve their issues themselves. Because there is nobody else who cares.

**Comment by 0xFFD** [comment]
This is still an issue, I have my own workspace setup script and I get stuck every time I try to install Spotify via chocolatey, I tried in VM which runs Windows 10 20H2 my current workspace uses Windows 10 22H2. 

Original poster is right when I remove Spotify Windows App, the installer works fine. To not encounter this issue, add this lines to your install script.

`Get-AppxPackage -AllUsers *Spotify* | Remove-AppxPackage                                # Uninstall Spotify`

Credits : https://github.com/geekygirlsarah/remove_win10_apps/blob/master/Uninstall_Win10_Apps.ps1

**Comment by pauby** [comment]
@0xFFD Can you run these two commands for me, both from Windows PowerShell, and provide the full output?

1. `Get-AppxProvisionedPackage -online | where displayname -like '*spotify*' | fl *`
2. `Get-AppxPackage | where name -like '*spotify*' | fl *`

I may try and fix the package based on the results.

**Comment by 0xFFD** [comment]
Already removed Spotify Windows Store version from my main lab using the command I posted earlier, and did the same with my VM's.  But for you I uninstalled my Spotify installation and downloaded it again via Windows Store here are the results ;

`Get-AppxProvisionedPackage -online | where displayname -like '*spotify*' | fl *`

Output is empty on this one, that may be because I didn't restarted my system after install.

`Get-AppxPackage | where name -like '*spotify*' | fl *`

Here is the output for 2nd command I ran powershell as admin. 

```
Name                   : SpotifyAB.SpotifyMusic
Publisher              : CN=453637B3-4E12-4CDF-B0D3-2A3C863BF6EF
PublisherId            : zpdnekdrzrea0
Architecture           : X64
ResourceId             :
Version                : 1.216.947.0
PackageFamilyName      : SpotifyAB.SpotifyMusic_zpdnekdrzrea0
PackageFullName        : SpotifyAB.SpotifyMusic_1.216.947.0_x64__zpdnekdrzrea0
InstallLocation        : C:\Program Files\WindowsApps\SpotifyAB.SpotifyMusic_1.216.947.0_x64__zpdnekdrzrea0
IsFramework            : False
PackageUserInformation : {}
IsResourcePackage      : False
IsBundle               : False
IsDevelopmentMode      : False
NonRemovable           : False
Dependencies           : {Microsoft.WindowsAppRuntime.1.2_2000.802.31.0_x64__8wekyb3d8bbwe,
                         Microsoft.VCLibs.140.00_14.0.32530.0_x64__8wekyb3d8bbwe}
IsPartiallyStaged      : False
SignatureKind          : Store
Status                 : Ok
```

I can provide my installation script if you want to take a look.

**Comment by pauby** [comment]
I think I have enough info. My question was really if Spotify installed itself as the user (`Get-AppxPackage`) or system-wide for each user (`Get-AppxProvisionedPackage`).

**Comment by 0xFFD** [comment]
@pauby (`Get-AppxProvisionedPackage`) returned nothing for me, this may be depending but I don't have any other Users in my lab only my Admin account. Installed Spotify directly from Windows Store with admin user.

**Comment by pauby** [comment]
No, I got that. I think it's because Spotify installs in the user context. Leave it with me.

**Comment by 0xFFD** [comment]
All good, if you need any other information/details I'll happily provide.

**Comment by pauby** [comment]
@0xFFD I've created a [PR for this](https://github.com/chocolatey-community/chocolatey-packages/pull/2263).
